abgreeve/tiny_stash
# Show items and trades directly in editor

## Prerequisites
1. A course with the Stash block added to it
2. The block_stash plugin checked out on the [new-webservices](https://github.com/abgreeve/moodle-block_stash/tree/new-webservices) branch
3. Tiny set as the default editor (or at least the editor for the user you use to test with)

## Testing
1. Create a page module in the course
2. In the "Content" field, click the stash icon:
    ![image](https://github.com/abgreeve/tiny_stash/assets/89116254/1b838fdb-2118-4850-8133-ff28de84a213)
4. Insert an item (either and existing item, or make a new one - it doesn't matter)
5. **Verify** the item preview inserts itself in to the editor
6. Press the "Save and display" button and **verify** the item is displayed
7. Edit the page
8. **Verify** the item shows in the "Content" field
9. Inspect the DB and **verify** no weird markup has been added (check the `mdl_page` table and look at the "content" column (**NB**: `<p>` tags are to be expected - but not any markup that looks like TinyMCE internal nodes)
10. Update the page, this time adding a trade in the "Content" field (via the stash button again)
11. **Verify** the trade inserts itself in to the editor
12. Press the "Save and display" button and **verify** the trade is displayed
13. Inspect the DB again and check for weirdness, **verify** there is none
14. Edit the page
15. **Verify** the trade shows in the "Content" field
16. Edit the page
17. Instead of using the insert functionality, insert a shortcode directly (as plain text)
18. Press the "Save and display" button and **verify** the item/trade displays
19. Edit the page
20. **Verify** the item/trade shows in the "Content" field
21. Inspect the DB again and **verify** no weird HTML crept in
22. Perform exploratory testing with different combinations of trades and items.
23. Bonus: Compare the behaviour to the plugin pre-patch. For example:
    1. (Pre patch): Put two shortcodes for items next to each other in the editor
    2. Press "Save and display", and observe how it looks. 
    3. (Post patch): Insert two items and put them next to each other in the editor
    5. Press "Save and display", and **verify** it appears the same as it did in 23.i

## Future considerations
It should be possible to use the values sent through to the editor from plugininfo.php to avoid further webservice calls. The variables `itemsData` and `tradeData` are shared between all editors on the page, as JavaScript modules are evaluated only once.

So all editor instances on the page are already sharing data, in a sense.

It should be possible to pre-populate the relevant variables, and append new items to them as they are added. As opposed to a webservice call to retreive all data and repopulate the variable entirely. This would mean the only webservice calls required would be calls to save data.
